bluefog/common/basics.py

```diff
@@ -116,6 +119,26 @@ def local_rank(self) -> int:
             raise ValueError("BlueFog has not been initialized; use bf.init().")
         return local_rank
 
+    def machine_rank(self) -> int:
+        """A function that returns the BlueFog rank of the machine.
+
+        Returns:
+            An integer scalar with the BlueFog rank of the machine.
+        """
+        # TODO(hhb) This only supports the homogenous environment now. Currently it assumes all
```

Comment:
Add the check like assert self.local_rank() == (self.rank() % self.local_size())
assert (self.size() % self.local_size()) == 0

to make sure, we currently only support homogeneous case..

BTW. there is is_homogenous function in this case.